File: Task4/src/Negation.scala
/**
 * Created by Шолохов on 16.11.2014.
 */
case class Negation(expression : Expression) extends Expression{
  override def evaluate(values: Map[String, Boolean]): Boolean = !expression.evaluate(values)

  override def equals(o: Any): Boolean = {
    o match {
      case n : Negation => expression.equals(n.expression)
      case _ => false
    }
  }

  override def toString() : String =  "(" + "!" + expression.toString + ")"
}

File: Task4/src/Negation.java
import java.util.HashMap;

public class Negation extends Statement {
    public final Statement child;

    public Negation(Statement child) {
        this.child = child;
    }

    protected boolean compareImpl(String[] patterns, Statement other) {
        if (other instanceof Negation) {
            return child.compareWithinContext(patterns, ((Negation)other).child);
        } else {
            return false;
        }
    }

    public Statement substitutePatterns(Statement[] to) {
        return new Negation(child.substitutePatterns(to));
    }

    public Statement substituteTerm(Expression where, Expression what) {
        return new Negation(child.substituteTerm(where, what));
    }

    public boolean estimate(HashMap<String, Boolean> values) {
        return !child.estimate(values);
    }

    public String toString() {
        return "!(" + child + ")";
    }
}
